pkg/server/message.go

```diff
@@ -236,18 +233,117 @@ func (t *TranslateServiceServer) UpdateMessages(
 		return nil, status.Errorf(codes.Internal, "")
 	}
 
-	// If updating the original messages and populateTranslations flag is true, update all translated messages as well.
-	if params.messages.Original && params.populateTranslations {
-		if err := t.populateTranslatedMessages(ctx, params.serviceID, params.messages, msgs); err != nil {
+	if params.messages.Original && len(allMessages) > 1 {
+		// TODO: optimize performance when populateTranslations param is true, currently saveMessages() will be called twice.
+
+		// find original messages where text has been altered then translate & update associated messages for all translations.
+		if err := t.updateAlteredMessages(ctx, params.serviceID, allMessages, params.messages); err != nil {
 			return nil, err
 		}
+
+		// if populateTranslations flag is true, update all translated messages as well.
+		if params.populateTranslations {
+			if err := t.populateTranslatedMessages(ctx, params.serviceID, params.messages, allMessages); err != nil {
+				return nil, err
+			}
+		}
 	}
 
 	return messagesToProto(params.messages), nil
 }
 
 // helpers
 
+// updateAlteredMessages finds original messages where text has been altered,
+// translates & updates associated messages for all translations.
+func (t *TranslateServiceServer) updateAlteredMessages(
+	ctx context.Context,
+	serviceID uuid.UUID,
+	allMessages []model.Messages,
+	newOriginalMessages *model.Messages,
+) error {
+	// return if only one language is present
+	if len(allMessages) == 1 {
+		return nil
+	}
+
+	var originalMessages *model.Messages
+
+	for i := range allMessages {
+		if allMessages[i].Original {
+			originalMessages = &allMessages[i]
+			break
+		}
+	}
```

Comment:
I am thinking about making `[]model.Messages `a type, and this could be one of its methods. 
On the upside cleaner code, as the downside, much refactoring is needed, almost everywhere...